include/lbann/layers/transform/binary_cross_entropy.hpp

```diff
@@ -0,0 +1,162 @@
+void fp_compute() {
+  //@todo: check that the order is not flipped 
+  const auto& predictions = get_prev_activations(0);
+  const auto& ground_truth = get_prev_activations(1);
+  auto& output_local = get_activations().Matrix();
+  // Local matrices
+  const Mat& predictions_local = predictions.LockedMatrix();
+  const Mat& ground_truth_local = ground_truth.LockedMatrix();
+  
+  // Matrix parameters
+  const int local_height = predictions_local.Height();
+  const int local_width = predictions_local.Width();
+
+  // Compute sum of cross entropy terms
+  #pragma omp parallel for collapse(2)
+  for (int col = 0; col < local_width; ++col) {
+    for (int row = 0; row < local_height; ++row) {
+      const DataType true_val = ground_truth_local(row, col);
+      const DataType pred_val = predictions_local(row, col);
+      //@todo: add check when the same implementation in objective function is deprecated
+      /*#ifdef LBANN_DEBUG
+      binary_cross_entropy_debug::check_entry(ground_truth.GlobalRow(row),
+                                              ground_truth.GlobalCol(col),
+                                              true_val,
+                                              pred_val);
+      #endif // LBANN_DEBUG*/
+      DataType sum = DataType(0);
+      if (true_val > DataType(0)) {
+        sum += - true_val * std::log(pred_val);
+      }
+      if (true_val < DataType(1)) {
+        sum += - (EvalType(1) - true_val) * std::log(EvalType(1) - pred_val);
+      }
+      output_local(row,col) = sum;
+    }
+  }
+
+}
+
+void bp_compute() {
+
+  //@todo: check that the order is not flipped 
+  const auto& predictions = get_prev_activations(0);
+  const auto& ground_truth = get_prev_activations(1);
+  // Local matrices
+  const Mat& predictions_local = predictions.LockedMatrix();
+  const Mat& ground_truth_local = ground_truth.LockedMatrix();
+  auto& gradient_local = get_error_signals().Matrix();
+
+  // Matrix parameters
+  const El::Int local_height = gradient_local.Height();
+  const El::Int local_width = gradient_local.Width();
+
+  // Compute gradient
+  #pragma omp parallel for collapse(2)
+  for (El::Int col = 0; col < local_width; ++col) {
+    for (El::Int row = 0; row < local_height; ++row) {
+      const DataType true_val = ground_truth_local(row, col);
+      const DataType pred_val = predictions_local(row, col);
+      /*#ifdef LBANN_DEBUG
+      binary_cross_entropy_debug::check_entry(ground_truth.GlobalRow(row),
+                                              ground_truth.GlobalCol(col),
+                                              true_val,
+                                              pred_val);
+      #endif // LBANN_DEBUG*/
+      DataType grad_val = DataType(0);
+      if (true_val > DataType(0)) {
+        grad_val += - true_val / pred_val;
+      }
+      if (true_val < DataType(1)) {
+        grad_val += (DataType(1) - true_val) / (DataType(1) - pred_val);
+      }
+      gradient_local(row, col) = grad_val;
```

Comment:
We should multiply this by the previous error signal to get the correct error signal:
dL/d(pred) = dL/d(output) * (-label/pred + (1-label)/(1-pred))

We should also compute the gradient w.r.t. the ground truth:
dL/d(label) = dL/d(output) * (-log(pred) + log(1-pred))